swupd/rename.go

```diff
@@ -0,0 +1,187 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// BSDIFFSIZE is smallest size of file to consider using bsdiff
+const BSDIFFSIZE = 200
+
+func renameDetection(manifest *Manifest, added []*File, removed []*File, c config) {
+	if len(added) == 0 || len(removed) == 0 {
+		return // nothing to rename
+	}
+	added = trimrenamed(added) // Make copies of input slices, tidy up whilst we are here
+	removed = trimrenamed(removed)
+	if err := fixupstatfields(removed, manifest, &c); err != nil {
+		panic(err)
+	}
+	// Handle pure renames first, don't need to worry about size. Should we skip zero size?
+	// just add call to trimsmall if so
+	// Sort by Hash. No need to have tiebreaker on name as the Manifest links by hash
+	// so 2 identical files being renamed, e.g. python2.6/foo and python3.7/bar being
+	// renamed to python2.7/foo and python3.7/bar have no concept which is the source for the
+	// python2.7/foo. This decision might need to change if the way renames are shown in
+	// the manifest changes.
+	sort.Slice(added, func(i, j int) bool {
+		return added[i].Hash < added[j].Hash
+	})
+	sort.Slice(removed, func(i, j int) bool {
+		return removed[i].Hash < removed[j].Hash
+	})
+	for ax, rx := 0, 0; ax < len(added) && rx < len(removed); {
+		af := added[ax]
+		rf := removed[rx]
+		switch {
+		case af.Hash < rf.Hash:
+			ax++
+		case af.Hash > rf.Hash:
+			rx++
+		default: // Equal hash, so link
+			linkrenamepair(af, rf)
+			ax++
+			rx++
+		}
+	}
+	// Link things with the same names skippingdigits
+	// First remove small files (not worth sending diff) and files which have
+	// exact match
+	added = trimsmall(trimrenamed(added), BSDIFFSIZE)     // Make it explicit we are doing two steps
+	removed = trimsmall(trimrenamed(removed), BSDIFFSIZE) // Can trivially make it one pass.
+	//generate the pairs of *File and short name
+	pa := makePairedNames(added)
+	pr := makePairedNames(removed)
+	// Merge where short names match
+	for ax, rx := 0, 0; ax < len(pa) && rx < len(pr); {
+		af := pa[ax]
+		rf := pr[rx]
+		switch {
+		case af.n < rf.n:
+			ax++
+		case af.n > rf.n:
+			rx++
+		default: // Equal truncated name
+			linkrenamepair(af.f, rf.f)
+			ax++
+			rx++
+		}
+	}
+}
+
+// linkrenamepair links two files together
+func linkrenamepair(a, b *File) {
+	a.DeltaPeer = b
+	b.DeltaPeer = a
+	a.Rename = true
+	b.Rename = true
+}
+
+// trimrenamed returns an slice which has had files with DeltaPeers purged
+func trimrenamed(a []*File) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Status == StatusGhosted {
+			// TODO. Do we allow ghosted files to be source files for renames?
+			// I think not, as we don't manage their lifetimes, so they could
+			// be deleted. The counter argument is that rename is only an optimization
```

Comment:
The counter argument is essentially my argument. The whole point of ghosted files is to allow renames with them

Reply:
OK, I will allow ghosted files as sources for renames. It makes the code shorter, even if it makes the code longer.